src/renderers/shared/fiber/ReactFiberUpdateQueue.js

```diff
@@ -371,24 +184,26 @@ function getStateFromUpdate(update, instance, prevState, props) {
   }
 }
 
-function beginUpdateQueue(
+function beginUpdateQueue<State>(
   current: Fiber | null,
   workInProgress: Fiber,
-  queue: UpdateQueue,
+  queue: UpdateQueue<State>,
   instance: any,
-  prevState: any,
   props: any,
   renderExpirationTime: ExpirationTime,
 ): any {
   if (current !== null && current.updateQueue === queue) {
```

Comment:
Unrelated for possible follow up: This, and similar code in the insert path, could probably be hoisted out into something that knows about Fibers so that the update queue can be decoupled from the fiber containing it.

Reply:
👍  I did that here: https://github.com/facebook/react/pull/10624/commits/9154198c3b188f6c4db2d9094d57f397bd3f2154 Held off in this PR until it's actually needed.